File: 2018/HW-3/p5/p5.go
/*
 *
 * In The Name of God
 *
 * +===============================================
 * | Author:        Parham Alvani <parham.alvani@gmail.com>
 * |
 * | Creation Date: 08-11-2018
 * |
 * | File Name:     p5.go
 * +===============================================
 */

package main

import "fmt"

func main() {
	var a1, b1 int
	var a2, b2 int

	fmt.Scanf("%d %d", &a1, &b1)
	fmt.Scanf("%d %d", &a2, &b2)

	// based on catalan numbers, these segments have 2 forms and following intersections:
	// ( ( ) ) x4
	// a1 a2 b1 b2
	// a1 a2 b2 b1
	// a2 a1 b2 b1
	// a2 a1 b1 b2
	// ( ) ( ) x2
	// a2 b2 a1 b1
	// a1 b1 a2 b2

	if a1 < a2 {
		if b1 < a2 { // a1 b1 a2 b2
			fmt.Println(0)
		} else {
			if b1 < b2 { // a1 a2 b1 b2
				fmt.Println(b1 - a2)
			} else { // a1 a2 b2 b1
				fmt.Println(b2 - a2)
			}
		}
	} else {
		if b2 < a1 { // a2 b2 a1 b1
			fmt.Println(0)
		} else {
			if b1 < b2 { // a2 a1 b1 b2
				fmt.Println(b1 - a1)
			} else { // a2 a1 b2 b1
				fmt.Println(b2 - a1)
			}
		}
	}
}

File: Fall-2018/Midterm/p5.c
/*
 * In The Name of God
 * =======================================
 * [] File Name : p5.c
 *
 * [] Creation Date : 26-11-2019
 *
 * [] Created By : Parham Alvani <parham.alvani@gmail.com>
 * =======================================
*/
/*
* Copyright (c)  2019 Parham Alvani.
*/

#include <stdio.h>

int T(int, int);

int main() {
  int x1, x2, n;

  scanf("%d %d %d", &x1, &x2, &n);

  int z = 0;
  for (int i = x1; i <= x2; i++) {
    z += T(n, i);
  }
  printf("%d\n", z);
}

int T(int n, int x) {
  if (n == 0) {
    return 1;
  }

  if (n == 1) {
    return x;
  }

  return x * T(n - 1, x) + T(n - 2, x);
}
